### ISSUE TITLE
DNS Fails several times a day

### ISSUE THREAD

**Issue by Slyke** [pi-hole/pi-hole#2352]
**In raising this issue, I confirm the following:** `{please fill the checkboxes, e.g: [X]}`

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [] The issue I am reporting can be *replicated*.
- [X] The issue I am reporting isn't a duplicate (see [FAQs](https://github.com/pi-hole/pi-hole/wiki/FAQs), [closed issues](https://github.com/pi-hole/pi-hole/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aclosed%20), and [open issues](https://github.com/pi-hole/pi-hole/issues)).

**How familiar are you with the the source code relevant to this issue?:**

`1`

---
**Expected behaviour:**

`DNS queries are successful`

**Actual behaviour:**

`This happens randomly at any time, but it happens on all devices (Android, Windows, Linux) at the same time. Basically when I try to navigate to a website (any website, google, youtube, facebook, github etc), I'll get the error 'dns_probe_finished_bad_config', 'dns_probe_finished_nxdomain ' or an error saying that the website's IP address couldn't be found.`

**Steps to reproduce:**

`I don't know how to reproduce it. I don't know how to fix it. I only know that it happens randomly, and that it fixes itself usually within 1-4 minutes`

**Debug token provided by [uploading `pihole -d` log](https://discourse.pi-hole.net/t/the-pihole-command-with-examples/738#debug):**

`teeuqexd65` - Immediately after being fixed.
`2a5b12io1x` - When the error is happening

**Troubleshooting undertaken, and/or other relevant information:**

`Waiting a few minutes, or restarting the Pi will fix this issue. This happened on the last version, and still happens the latest version (released a few days ago).`

**Comment by Mcat12** [comment]
There's very little difference between the two debug logs. Check the dnsmasq log (`/var/log/pihole.log` or `pihole -t`) when there are issues and share the log snippet. If there's an issue on Pi-hole's side it should show up there.

**Comment by Slyke** [comment]
I left my ssh running for a few days. It just happened again.
```
pi@pidns:~ $ pihole -t
  [i] Press Ctrl-C to exit
14:51:01 dnsmasq[784]: 34 127.0.0.1/56893 query[A] api.github.com from 127.0.0.1
14:51:01 dnsmasq[784]: 34 127.0.0.1/56893 forwarded api.github.com to 8.8.4.4
14:51:01 dnsmasq[784]: 34 127.0.0.1/56893 forwarded api.github.com to 8.8.8.8
14:51:01 dnsmasq[784]: 35 127.0.0.1/56893 query[AAAA] api.github.com from 127.0.0.1
14:51:01 dnsmasq[784]: 35 127.0.0.1/56893 forwarded api.github.com to 8.8.4.4
14:51:01 dnsmasq[784]: 35 127.0.0.1/56893 reply api.github.com is NODATA-IPv6
14:51:01 dnsmasq[784]: 34 127.0.0.1/56893 reply api.github.com is 192.30.253.116
14:51:01 dnsmasq[784]: 34 127.0.0.1/56893 reply api.github.com is 192.30.253.117
14:51:01 dnsmasq[784]: 36 127.0.0.1/55620 query[A] api.github.com from 127.0.0.1
14:51:01 dnsmasq[784]: 36 127.0.0.1/55620 cached api.github.com is 192.30.253.117
14:51:01 dnsmasq[784]: 36 127.0.0.1/55620 cached api.github.com is 192.30.253.116
14:51:01 dnsmasq[784]: 37 127.0.0.1/55620 query[AAAA] api.github.com from 127.0.0.1
14:51:01 dnsmasq[784]: 37 127.0.0.1/55620 cached api.github.com is NODATA-IPv6
14:51:02 dnsmasq[784]: 38 127.0.0.1/45456 query[A] api.github.com from 127.0.0.1
14:51:02 dnsmasq[784]: 38 127.0.0.1/45456 cached api.github.com is 192.30.253.116
14:51:02 dnsmasq[784]: 38 127.0.0.1/45456 cached api.github.com is 192.30.253.117
14:51:02 dnsmasq[784]: 39 127.0.0.1/45456 query[AAAA] api.github.com from 127.0.0.1
14:51:02 dnsmasq[784]: 39 127.0.0.1/45456 cached api.github.com is NODATA-IPv6
tail: /var/log/pihole.log: file truncated
Aug 15 14:51:01 dnsmasq[784]: 40 127.0.0.1/34474 query[A] api.github.com from 127.0.0.1
Aug 15 14:51:01 dnsmasq[784]: 40 127.0.0.1/34474 forwarded api.github.com to 8.8.4.4
Aug 15 14:51:01 dnsmasq[784]: 40 127.0.0.1/34474 forwarded api.github.com to 8.8.8.8
Aug 15 14:51:01 dnsmasq[784]: 41 127.0.0.1/34474 query[AAAA] api.github.com from 127.0.0.1
Aug 15 14:51:01 dnsmasq[784]: 41 127.0.0.1/34474 forwarded api.github.com to 8.8.4.4
Aug 15 14:51:01 dnsmasq[784]: 41 127.0.0.1/34474 reply api.github.com is NODATA-IPv6
Aug 15 14:51:01 dnsmasq[784]: 40 127.0.0.1/34474 reply api.github.com is 192.30.253.116
Aug 15 14:51:01 dnsmasq[784]: 40 127.0.0.1/34474 reply api.github.com is 192.30.253.117
Aug 15 14:51:02 dnsmasq[784]: 42 127.0.0.1/37764 query[A] api.github.com from 127.0.0.1
Aug 15 14:51:02 dnsmasq[784]: 42 127.0.0.1/37764 cached api.github.com is 192.30.253.117
Aug 15 14:51:02 dnsmasq[784]: 42 127.0.0.1/37764 cached api.github.com is 192.30.253.116
Aug 15 14:51:02 dnsmasq[784]: 43 127.0.0.1/37764 query[AAAA] api.github.com from 127.0.0.1
Aug 15 14:51:02 dnsmasq[784]: 43 127.0.0.1/37764 cached api.github.com is NODATA-IPv6
Aug 15 14:51:02 dnsmasq[784]: 44 127.0.0.1/41236 query[A] api.github.com from 127.0.0.1
Aug 15 14:51:02 dnsmasq[784]: 44 127.0.0.1/41236 cached api.github.com is 192.30.253.116
Aug 15 14:51:02 dnsmasq[784]: 44 127.0.0.1/41236 cached api.github.com is 192.30.253.117
Aug 15 14:51:02 dnsmasq[784]: 45 127.0.0.1/41236 query[AAAA] api.github.com from 127.0.0.1
Aug 15 14:51:02 dnsmasq[784]: 45 127.0.0.1/41236 cached api.github.com is NODATA-IPv6
```
Happened after the final entry above `^^^`

And here is `tail -n 60 /var/log/pihole.log` of
```
pi@pidns:~ $ tail -n 60 /var/log/pihole.log
Aug 15 14:51:01 dnsmasq[784]: 40 127.0.0.1/34474 query[A] api.github.com from 127.0.0.1
Aug 15 14:51:01 dnsmasq[784]: 40 127.0.0.1/34474 forwarded api.github.com to 8.8.4.4
Aug 15 14:51:01 dnsmasq[784]: 40 127.0.0.1/34474 forwarded api.github.com to 8.8.8.8
Aug 15 14:51:01 dnsmasq[784]: 41 127.0.0.1/34474 query[AAAA] api.github.com from 127.0.0.1
Aug 15 14:51:01 dnsmasq[784]: 41 127.0.0.1/34474 forwarded api.github.com to 8.8.4.4
Aug 15 14:51:01 dnsmasq[784]: 41 127.0.0.1/34474 reply api.github.com is NODATA-IPv6
Aug 15 14:51:01 dnsmasq[784]: 40 127.0.0.1/34474 reply api.github.com is 192.30.253.116
Aug 15 14:51:01 dnsmasq[784]: 40 127.0.0.1/34474 reply api.github.com is 192.30.253.117
Aug 15 14:51:02 dnsmasq[784]: 42 127.0.0.1/37764 query[A] api.github.com from 127.0.0.1
Aug 15 14:51:02 dnsmasq[784]: 42 127.0.0.1/37764 cached api.github.com is 192.30.253.117
Aug 15 14:51:02 dnsmasq[784]: 42 127.0.0.1/37764 cached api.github.com is 192.30.253.116
Aug 15 14:51:02 dnsmasq[784]: 43 127.0.0.1/37764 query[AAAA] api.github.com from 127.0.0.1
Aug 15 14:51:02 dnsmasq[784]: 43 127.0.0.1/37764 cached api.github.com is NODATA-IPv6
Aug 15 14:51:02 dnsmasq[784]: 44 127.0.0.1/41236 query[A] api.github.com from 127.0.0.1
Aug 15 14:51:02 dnsmasq[784]: 44 127.0.0.1/41236 cached api.github.com is 192.30.253.116
Aug 15 14:51:02 dnsmasq[784]: 44 127.0.0.1/41236 cached api.github.com is 192.30.253.117
Aug 15 14:51:02 dnsmasq[784]: 45 127.0.0.1/41236 query[AAAA] api.github.com from 127.0.0.1
Aug 15 14:51:02 dnsmasq[784]: 45 127.0.0.1/41236 cached api.github.com is NODATA-IPv6
```

Funny thing is that I went to youtube when this happened, not github, and I've been to many sites over the last few days, not just github like these logs indicate.

**Comment by Slyke** [comment]
It just happened again. Seems the log files were cleared this time:
```
pi@pidns:~ $ tail -n 60 /var/log/pihole.log
pi@pidns:~ $ uptime
 01:49:41 up 2 days, 20:49,  1 user,  load average: 0.00, 0.00, 0.00
```

I checked the syslogs though and noticed that it checked for an update right before it happened:
```
Aug 16 00:09:44 pidns systemd[1]: Starting Clean php session files...
Aug 16 00:09:44 pidns systemd[1]: Started Clean php session files.
Aug 16 00:10:01 pidns CRON[18553]: (root) CMD (   PATH="$PATH:/usr/local/bin/" pihole updatechecker local)
Aug 16 00:17:01 pidns CRON[18574]: (root) CMD (   cd / && run-parts --report /etc/cron.hourly)
Aug 16 00:20:01 pidns CRON[18589]: (root) CMD (   PATH="$PATH:/usr/local/bin/" pihole updatechecker local)
Aug 16 00:30:01 pidns CRON[18611]: (root) CMD (   PATH="$PATH:/usr/local/bin/" pihole updatechecker local)
Aug 16 00:39:01 pidns CRON[18633]: (root) CMD (  [ -x /usr/lib/php/sessionclean ] && if [ ! -d /run/systemd/system ]; then /usr/lib/php/sessionclean; fi)
Aug 16 00:39:44 pidns systemd[1]: Starting Clean php session files...
Aug 16 00:39:44 pidns systemd[1]: Started Clean php session files.
Aug 16 00:40:01 pidns CRON[18685]: (root) CMD (   PATH="$PATH:/usr/local/bin/" pihole updatechecker local)
Aug 16 00:50:01 pidns CRON[18706]: (root) CMD (   PATH="$PATH:/usr/local/bin/" pihole updatechecker local)
Aug 16 01:00:01 pidns CRON[18728]: (root) CMD (   PATH="$PATH:/usr/local/bin/" pihole updatechecker local)
Aug 16 01:09:01 pidns CRON[18750]: (root) CMD (  [ -x /usr/lib/php/sessionclean ] && if [ ! -d /run/systemd/system ]; then /usr/lib/php/sessionclean; fi)
Aug 16 01:09:44 pidns systemd[1]: Starting Clean php session files...
Aug 16 01:09:44 pidns systemd[1]: Started Clean php session files.
Aug 16 01:10:02 pidns CRON[18802]: (root) CMD (   PATH="$PATH:/usr/local/bin/" pihole updatechecker local)
Aug 16 01:17:01 pidns CRON[18824]: (root) CMD (   cd / && run-parts --report /etc/cron.hourly)
Aug 16 01:20:01 pidns CRON[18838]: (root) CMD (   PATH="$PATH:/usr/local/bin/" pihole updatechecker local)
Aug 16 01:30:01 pidns CRON[18860]: (root) CMD (   PATH="$PATH:/usr/local/bin/" pihole updatechecker local)
Aug 16 01:39:01 pidns CRON[18882]: (root) CMD (  [ -x /usr/lib/php/sessionclean ] && if [ ! -d /run/systemd/system ]; then /usr/lib/php/sessionclean; fi)
Aug 16 01:39:44 pidns systemd[1]: Starting Clean php session files...
Aug 16 01:39:44 pidns systemd[1]: Started Clean php session files.
Aug 16 01:40:01 pidns CRON[18934]: (root) CMD (   PATH="$PATH:/usr/local/bin/" pihole updatechecker local)
Aug 16 01:50:01 pidns CRON[18978]: (root) CMD (   PATH="$PATH:/usr/local/bin/" pihole updatechecker local)
pi@pidns:~ $ date
Thu 16 Aug 01:50:35 UTC 2018
```

Possible that the update check is causing it?

**Comment by Slyke** [comment]
I think these logs may have found the issue. After happening again, I checked syslog and found this:
```
Aug 16 04:39:44 pidns systemd[1]: Starting Clean php session files...
Aug 16 04:39:44 pidns systemd[1]: Started Clean php session files.
Aug 16 04:40:01 pidns CRON[19700]: (root) CMD (   PATH="$PATH:/usr/local/bin/" pihole updatechecker local)
Aug 16 04:50:01 pidns CRON[19722]: (root) CMD (   PATH="$PATH:/usr/local/bin/" pihole updatechecker local)
Aug 16 05:00:01 pidns CRON[19744]: (root) CMD (   PATH="$PATH:/usr/local/bin/" pihole updatechecker local)
Aug 16 05:00:22 pidns dhcpcd[576]: tun0: adding default route via 192.168.2.1
Aug 16 05:00:22 pidns dhcpcd[576]: if_route (ADD): Network is unreachable
Aug 16 05:00:22 pidns dhcpcd[576]: tun0: adding default route via 192.168.2.1
Aug 16 05:00:22 pidns dhcpcd[576]: if_route (ADD): Network is unreachable
pi@pidns:~ $ date
Thu 16 Aug 05:01:09 UTC 2018
```

Having the network unreachable could possible do it? Looks like it was caused by the update checker again. Maybe the raspberry pi is having power issues?

**Comment by admiralsmaster** [comment]
Same goes for me. Sometimes the IP resolution returns NO-IP. After a few minutes waiting time, the resolution returns an IP. 

I have a pihole with a local unbound DNS upstream server.

**Comment by Slyke** [comment]
Nope. I've got one raspberry pi with the DNS entries set to 8.8.8.8 and 8.8.4.4 and it never has any issues. I don't see the network go down (I can be streaming a game to youtube, open up Steam overlay to search for something, and I get an error page due to no domain resolution, without the stream dying).

I suspect that the pi is having power issues, and have currently ordered a usb power monitor from Amazon (waiting for delivery). I think the update script is causing the CPU to use more power and causing brown outs, but it's a shot in the dark.

**Comment by admiralsmaster** [comment]
I've been running my pihole for over a year. I don't see this issue before updating to version 4.
Next time I check my Internet access by calling an IP directly.

**Comment by Slyke** [comment]
My power monitor arrived today. Going to plug it in and watch. In the mean time, got some more logs when it happened

```
Aug 18 01:39:44 pidns systemd[1]: Starting Clean php session files...
Aug 18 01:39:44 pidns systemd[1]: Started Clean php session files.
Aug 18 01:40:01 pidns CRON[31727]: (root) CMD (   PATH="$PATH:/usr/local/bin/" pihole updatechecker local)
Aug 18 01:50:01 pidns CRON[31749]: (root) CMD (   PATH="$PATH:/usr/local/bin/" pihole updatechecker local)
Aug 18 02:00:02 pidns CRON[31771]: (root) CMD (   PATH="$PATH:/usr/local/bin/" pihole updatechecker local)
Aug 18 02:09:01 pidns CRON[31793]: (root) CMD (  [ -x /usr/lib/php/sessionclean ] && if [ ! -d /run/systemd/system ]; then /usr/lib/php/sessionclean; fi)
Aug 18 02:09:44 pidns systemd[1]: Starting Clean php session files...
Aug 18 02:09:44 pidns systemd[1]: Started Clean php session files.
Aug 18 02:10:01 pidns CRON[31845]: (root) CMD (   PATH="$PATH:/usr/local/bin/" pihole updatechecker local)
Aug 18 02:17:01 pidns CRON[31866]: (root) CMD (   cd / && run-parts --report /etc/cron.hourly)
Aug 18 02:20:01 pidns CRON[31881]: (root) CMD (   PATH="$PATH:/usr/local/bin/" pihole updatechecker local)
Aug 18 02:30:01 pidns CRON[31903]: (root) CMD (   PATH="$PATH:/usr/local/bin/" pihole updatechecker local)
Aug 18 02:34:21 pidns systemd[1]: Created slice User Slice of pi.
Aug 18 02:34:21 pidns systemd[1]: Starting User Manager for UID 1000...
Aug 18 02:34:21 pidns systemd[1]: Started Session c13 of user pi.
Aug 18 02:34:21 pidns systemd[31926]: Listening on GnuPG cryptographic agent and passphrase cache (restricted).
Aug 18 02:34:21 pidns systemd[31926]: Listening on GnuPG cryptographic agent and passphrase cache.
Aug 18 02:34:21 pidns systemd[31926]: Reached target Timers.
Aug 18 02:34:21 pidns systemd[31926]: Reached target Paths.
Aug 18 02:34:21 pidns systemd[31926]: Listening on GnuPG cryptographic agent (access for web browsers).
Aug 18 02:34:21 pidns systemd[31926]: Listening on GnuPG cryptographic agent (ssh-agent emulation).
Aug 18 02:34:21 pidns systemd[31926]: Reached target Sockets.
Aug 18 02:34:21 pidns systemd[31926]: Reached target Basic System.
Aug 18 02:34:21 pidns systemd[31926]: Reached target Default.
Aug 18 02:34:21 pidns systemd[31926]: Startup finished in 39ms.
Aug 18 02:34:21 pidns systemd[1]: Started User Manager for UID 1000.
pi@pidns:~ $ date
Sat 18 Aug 02:34:30 UTC 2018
pi@pidns:~ $ uptime
 02:35:23 up 4 days, 21:35,  1 user,  load average: 0.00, 0.00, 0.00
```

Again it seems to of happened immediately after checking for an update.

**Comment by Slyke** [comment]
![image](https://user-images.githubusercontent.com/6442613/44305818-b3dab980-a34f-11e8-9af8-b50f29b7a719.png)
So during another one of these DNS dropout issues, I had `ping`, `top` and `nslookup` all running. It seems that ping and nslookup is still successful on the pi, but on no other machines on the network that are using pihole. It also dropped out immediately before pihole ran the update script.

I also looked at the voltage and amps the pi was pulling. It was pulling about 5.35v and 0.35a which seems okay. The voltage didn't drop below 5.2v. There wasn't any brownouts.

**Comment by admiralsmaster** [comment]
I change from the local unbound upstream resolver bach to Quad9.
Now the issue stops. At my pihole the usage of unbound

**Comment by Mcat12** [comment]
@Slyke the kernel is complaining of under-voltage, and the `top` output does not show the programs that are running. I can't see anything pointing to Pi-hole as the cause of this, so it might just be a strange interaction among your devices, programs, and Pi-hole.

**Comment by Slyke** [comment]
The low voltage is coming because I'm monitoring the power with a USB device. If I remove it, it goes away. I'm going to start monitoring if dnsmasq service is up. The only thing I can gather from these logs is that it always happens immediately before the update script runs.

**Comment by Slyke** [comment]
I've turned off Pihole for now. I couldn't figure out what was causing it and it became too annoying. It could happen several times an hour, then behave for days.

**Comment by tayyyTF** [comment]
I have been having similar issues. Pihole will randomly stop working after a few hours, and no clients can resolve anything. Even on the pi itself, resolution doesn't work:

`; <<>> DiG 9.10.3-P4-Raspbian <<>> google.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOTAUTH, id: 21602
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;google.com.                    IN      A

;; Query time: 13 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Sun Sep 09 13:22:16 EDT 2018
;; MSG SIZE  rcvd: 39`

I have to edit `/etc/resolv.conf` to change the local resolution to an upstream one or I cant even apt update/upgrade. Never had a problem before the recent update.

**Comment by Mcat12** [comment]
@tayyyTF Run `pihole -d` for a debug token. Do the upstream servers FTL is using work if you use them directly instead of through FTL's dnsmasq?

**Comment by tayyyTF** [comment]
Yes the upstream work directly, while using pihole does not work:

`pi@raspberrypi:~ $ cat /etc/resolv.conf`
`# Dynamic resolv.conf(5) file for glibc resolver(3) generated by resolvconf(8)`
`#     DO NOT EDIT THIS FILE BY HAND -- YOUR CHANGES WILL BE OVERWRITTEN`
`nameserver 127.0.0.1`

===============================================

`pi@raspberrypi:~ $ dig google.com`

`; <<>> DiG 9.10.3-P4-Raspbian <<>> google.com`
`;; global options: +cmd`
`;; Got answer:`
`;; ->>HEADER<<- opcode: QUERY, status: NOTAUTH, id: 16842`
`;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1`

`;; OPT PSEUDOSECTION:`
`; EDNS: version: 0, flags:; udp: 4096`
`;; QUESTION SECTION:`
`;google.com.                    IN      A`

`;; Query time: 11 msec`
`;; SERVER: 127.0.0.1#53(127.0.0.1)`
`;; WHEN: Tue Sep 18 11:05:28 EDT 2018`
`;; MSG SIZE  rcvd: 39`

========================================================

24.226.1.93 is one of the dns servers that I am using.

`pi@raspberrypi:~ $ dig google.com @24.226.1.93`

`; <<>> DiG 9.10.3-P4-Raspbian <<>> google.com @24.226.1.93`
`;; global options: +cmd`
`;; Got answer:`
`;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 35429`
`;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1`

`;; OPT PSEUDOSECTION:`
`; EDNS: version: 0, flags:; udp: 4096`
`;; QUESTION SECTION:`
`;google.com.                    IN      A`

`;; ANSWER SECTION:`
`google.com.             916     IN      A       172.217.1.174`

`;; Query time: 9 msec`
`;; SERVER: 24.226.1.93#53(24.226.1.93)`
`;; WHEN: Tue Sep 18 11:08:31 EDT 2018`
`;; MSG SIZE  rcvd: 55`


======================

Here is the output of pihole -d: 

[piholelog.txt](https://github.com/pi-hole/pi-hole/files/2393300/piholelog.txt)

**Comment by Mcat12** [comment]
I don't see any issues in your debug log. When you send Pi-hole a DNS request that fails, share the lines it generates in `/var/log/pihole.log`

**Comment by Slyke** [comment]
@tayyyTF can you try with `8.8.8.8`? That's what my Pihole was using, but this issue was still happening.

**Comment by tayyyTF** [comment]
@Slyke My current working DNS settings are below. I have both of Google's, both of Cloudflare's, both of Level3's, and my ISP's DNS server selected.

![image](https://user-images.githubusercontent.com/10136313/45783985-5af38f00-bc35-11e8-8269-afa5695c5633.png)

My ISP's DNS servers respond much quicker than the third party ones, so PiHole was sending the majority of my requests to them. Unfortunately, one of their DNS servers was either misconfigured or is no longer running because its IP was responding with a NOAUTH error for almost every request it received. When I removed that custom DNS server from my list, PiHole returned to its proper functionality. I can't say I have experienced any problems with `8.8.8.8`.
